corehq/ex-submodules/toggle/models.py

```diff
@@ -26,7 +29,10 @@ def save(self, **params):
         if ('_id' not in self._doc):
             self._doc['_id'] = generate_toggle_id(self.slug)
         self.last_modified = datetime.utcnow()
+        old_doc_rev = self._rev
         super(Toggle, self).save(**params)
+        sql_save_with_resource_conflict(SqlToggle, self, old_doc_rev)
```

Comment:
If the SQL save fails with `ResourceConflict` will it prevent future saves from succeeding because the revs in couch and SQL are out of sync?

Reply:
Hmm. that is frustrating and it appears to be something that we overlooked.

For this migration, and other small migrations, I could just skip straight to not saving to couch and only saving to SQL, but I really like the reversion path to be a simple "revert this PR".

Any ideas you have to figure this one out?

Reply:
A few ideas, not sure any of them is bulletproof, but thought I'd maybe re-kickstart the conversation

What if while this is going on we still treat CouchDB as the source of truth, and if SQL fails with a `ResourceConflict` then we just pull whatever is in Couch and overwrite what's in SQL with that?

Or alternatively, we could do some "locking" around these two lines so you're guaranteed that writes happen in pairs. The failure would be if postgres is down but couch is not then the write would never make it to SQL.

A combination of the two may get us pretty close

Reply:
I was hoping to have these in separate phases:

1. treat couch as source of truth, sql is a replica
2. treat sql as source of truth, couch is a replica

#1 is important as it allows us to migrate in the background
#2 is important as it allows us to revert back to couch if we find a bug with the process at some point.

An issue with that is that couch needs revisions and there's no way to set the revision without saving it to couch. We could do some form of locking, but that does increase complexity of saving.

For feature flags, we could just do it all in one go, meaning migrate all feature flags to SQL and stop using couch though we couldn't revert (or would lose data)

A good, general solution might be to just never save revision to SQL, or slightly modifiying this process to not use `DocumentField` (or only use it when document attributes are complex) and move the wrapping to a document outside of the field. There should be a way to incorporate https://github.com/dimagi/dimagi-utils/pull/272/ into the process and since Giovanni wrote it, I'm sure its solid and better than whatever I'd write

Reply:
re: There should be a way to incorporate dimagi/dimagi-utils#272 into the process...

I hadn't seen that before. Thanks for sharing. Here are a few key differences I noticed between that and the approach in this PR:

- It [maps individual fields in Couch to fields in SQL](https://github.com/dimagi/dimagi-utils/blob/f8ae901669fc9acd9387ed9a4b858d7c3593c130/dimagi/utils/couch/migration.py#L35-L41) rather than saving the [entire couch doc in a single SQL field](https://github.com/dimagi/commcare-hq/blob/9c3e516e70f52174f51b93eea2eb3716d8dd7f11/corehq/ex-submodules/toggle/models.py#L95). Each has pros and cons, but I don't think either affects data integrity in a critical way.
- It does not handle concurrent edits with conflicts in a robust way. If there is a `ResourceConflict` or any other error during the sync/save it simply [logs the error and continues as if there was no problem](https://github.com/dimagi/dimagi-utils/blob/f8ae901669fc9acd9387ed9a4b858d7c3593c130/dimagi/utils/couch/migration.py#L132-L137), which means Couch and SQL could get out of sync and then a later save would overwrite the differences as if there had never been a conflict. This is quite optimistic, and a fine policy to follow if data loss is not a concern.

I think the approach in this PR is more robust since it actually attempts to deal with concurrent edits in a way that can be responded to and properly resolved in code rather than requiring someone to keep a very close eye on Sentry (and pray that no data ever gets lost that actually matters).